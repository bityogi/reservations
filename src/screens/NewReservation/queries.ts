import gql from 'graphql-tag';

import { client } from '../../util/requests';
import { reservationQuery } from '../../util/queries';
import { newReservation } from '../../util/mutations';

export type NewReservationInputType = {
  name: string,
  hotelName: string,
  arrivalDate: string,
  departureDate: string
}

export const addNewReservation = async (input: NewReservationInputType) => {

  const variables = {
    input
  }
  
  const { data: { reservation } } = await client.mutate({ 
    mutation: newReservation,
    variables,
    update: (cache, { data }) => {
      cache.writeQuery({
        query: reservationQuery,
        variables: {
          where: {
            id: data.reservation.id
          }
        },
        data
      })
    }
  });

  return reservation;
} 