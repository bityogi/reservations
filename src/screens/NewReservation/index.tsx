import React, { PureComponent } from 'react';
import {
  View,
  Text,
  StyleSheet,
  TouchableHighlight,
  TextInput,
} from 'react-native';
import DateTimePicker from 'react-native-modal-datetime-picker';

import { formatDate } from '../../util';
import { NavigationProp } from '../../util/types';
import { addNewReservation } from './queries';

const styles = StyleSheet.create({
  container: {
    flex: 1,
  },
  fieldContainer: {
    marginTop: 20,
    marginBottom: 20,
    backgroundColor: '#fff',
  },
  text: {
    height: 40,
    margin: 0,
    marginRight: 7,
    paddingLeft: 10,
  },
  button: {
    height: 50,
    backgroundColor: '#48BBEC',
    borderColor: '#48BBEC',
    alignSelf: 'stretch',
    margin: 10,
    justifyContent: 'center',
    alignItems: 'center',
    borderRadius: 5,
  },
  buttonText: {
    color: '#fff',
    fontSize: 18,
  },
  borderTop: {
    borderColor: '#edeeef',
    borderTopWidth: 0.5,
  },
});

interface NewReservationState {
  name: string,
  hotelName: string,
  arrivalDate: Date,
  departureDate: Date,
  showArrivalDatePicker: boolean,
  showDepartureDatePicker: boolean
}

class NewReservation extends PureComponent<NavigationProp, NewReservationState> {
  static navigationOptions = {
    title: 'New Reservation'
  };

  state = {
    name: '',
    hotelName: '',
    arrivalDate: new Date(),
    departureDate: new Date(),
    showArrivalDatePicker: false,
    showDepartureDatePicker: false,
  }

  handlePress = () => {
    const { 
      name, 
      hotelName, 
      arrivalDate, 
      departureDate 
    } = this.state;
    
    const inputValues = {
      name,
      hotelName,
      arrivalDate: formatDate(arrivalDate),
      departureDate: formatDate(departureDate)
    }
    addNewReservation(inputValues)
      .then(reservation => {
        const { navigation } = this.props;
        navigation.navigate('ReservationDetail', {
          reservationId: reservation.id,
          refresh: true
        })
      })
      .catch(error => {
        console.error('Error occured while adding new reservation: ', error);
      });
    
  }

  handleArrivalDatePicked = (date: Date) => {
    this.setState({ 
      arrivalDate: date,
      showArrivalDatePicker: false,
    });
  }

  handleDepartureDatePicked = (date: Date) => {
    this.setState({
      departureDate: date,
      showDepartureDatePicker: false,
    })
  }

  render() {
    const { 
      name, 
      hotelName, 
      arrivalDate, 
      departureDate,
      showArrivalDatePicker,
      showDepartureDatePicker,
    } = this.state;

    const isValid = (name && hotelName && arrivalDate && departureDate);

    return (
      <View style={styles.container}>
        <View style={styles.fieldContainer}>
          <TextInput 
            style={styles.text}
            placeholder='Name on Reservation'
            spellCheck={false}
            onChangeText={(value) => this.setState({ name: value })}
            value={name}
          />
          <TextInput 
            style={[styles.text, styles.borderTop]}
            placeholder='Hotel Name'
            spellCheck={false}
            onChangeText={(value) => this.setState({ hotelName: value })}
            value={hotelName}
          />
          <TextInput 
            style={[styles.text, styles.borderTop]}
            placeholder='Arrival Date'
            spellCheck={false}
            onChangeText={(value) => this.setState({ arrivalDate: new Date(value) })}
            value={formatDate(arrivalDate)}
            editable={!showArrivalDatePicker}
            onFocus={() => this.setState({ showArrivalDatePicker: true })}
          />
          <TextInput 
            style={[styles.text, styles.borderTop]}
            placeholder='Departure Date'
            spellCheck={false}
            onChangeText={(value) => this.setState({ departureDate: new Date(value) })}
            value={formatDate(departureDate)}
            editable={!showDepartureDatePicker}
            onFocus={() => this.setState({ showDepartureDatePicker: true })}
          />
          <DateTimePicker
            isVisible={showArrivalDatePicker}
            mode="date"
            onConfirm={this.handleArrivalDatePicked}
            onCancel={() => this.setState({ showArrivalDatePicker: false })}
          />
          <DateTimePicker
            isVisible={showDepartureDatePicker}
            mode="date"
            onConfirm={this.handleDepartureDatePicked}
            onCancel={() => this.setState({ showDepartureDatePicker: false })}
          />
        </View>
        <TouchableHighlight
          onPress={this.handlePress}
          style={styles.button}
          disabled={!isValid}
        >
          <Text style={styles.buttonText}>New Reservation</Text>
        </TouchableHighlight>
        
      </View>
      
    )
  }
}

export default NewReservation;