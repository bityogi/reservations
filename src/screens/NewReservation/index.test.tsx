import 'react-native';
import React from 'react';
// Note: test renderer must be required after react-native.
import { shallow } from 'enzyme';
import NewReservation from './index';

jest.mock("NativeModules", () => ({
  UIManager: {
    RCTView: () => ({
      directEventTypes: {}
    })
  },
  KeyboardObserver: {},
  RNGestureHandlerModule: {
    attachGestureHandler: jest.fn(),
    createGestureHandler: jest.fn(),
    dropGestureHandler: jest.fn(),
    updateGestureHandler: jest.fn(),
    State: {},
    Directions: {}
  }
}));


jest.mock('../../util/requests');
jest.mock('./queries');



const navigationProp = {
    navigate: (to: string, options?: object) => { console.log('Navigating') },
    getParam: (param: string) => { return '1' },
    addListener: (listenEvent: string, payload: object) => {}
}

describe('NewReservation', () => {
  it('renders correctly', () => {
    const component = shallow(<NewReservation navigation={navigationProp} />);
    expect(component).toMatchSnapshot();
  })
})