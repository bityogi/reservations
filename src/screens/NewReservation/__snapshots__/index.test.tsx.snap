// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NewReservation renders correctly 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "backgroundColor": "#fff",
        "marginBottom": 20,
        "marginTop": 20,
      }
    }
  >
    <TextInput
      allowFontScaling={true}
      onChangeText={[Function]}
      placeholder="Name on Reservation"
      rejectResponderTermination={true}
      spellCheck={false}
      style={
        Object {
          "height": 40,
          "margin": 0,
          "marginRight": 7,
          "paddingLeft": 10,
        }
      }
      underlineColorAndroid="transparent"
      value=""
    />
    <TextInput
      allowFontScaling={true}
      onChangeText={[Function]}
      placeholder="Hotel Name"
      rejectResponderTermination={true}
      spellCheck={false}
      style={
        Array [
          Object {
            "height": 40,
            "margin": 0,
            "marginRight": 7,
            "paddingLeft": 10,
          },
          Object {
            "borderColor": "#edeeef",
            "borderTopWidth": 0.5,
          },
        ]
      }
      underlineColorAndroid="transparent"
      value=""
    />
    <TextInput
      allowFontScaling={true}
      editable={true}
      onChangeText={[Function]}
      onFocus={[Function]}
      placeholder="Arrival Date"
      rejectResponderTermination={true}
      spellCheck={false}
      style={
        Array [
          Object {
            "height": 40,
            "margin": 0,
            "marginRight": 7,
            "paddingLeft": 10,
          },
          Object {
            "borderColor": "#edeeef",
            "borderTopWidth": 0.5,
          },
        ]
      }
      underlineColorAndroid="transparent"
      value="10 Oct 2020"
    />
    <TextInput
      allowFontScaling={true}
      editable={true}
      onChangeText={[Function]}
      onFocus={[Function]}
      placeholder="Departure Date"
      rejectResponderTermination={true}
      spellCheck={false}
      style={
        Array [
          Object {
            "height": 40,
            "margin": 0,
            "marginRight": 7,
            "paddingLeft": 10,
          },
          Object {
            "borderColor": "#edeeef",
            "borderTopWidth": 0.5,
          },
        ]
      }
      underlineColorAndroid="transparent"
      value="20 Oct 2020"
    />
    <CustomDatePickerIOS
      cancelTextIOS="Cancel"
      confirmTextIOS="Confirm"
      date={2019-08-13T14:43:38.802Z}
      dismissOnBackdropPressIOS={true}
      hideTitleContainerIOS={false}
      isVisible={false}
      mode="date"
      neverDisableConfirmIOS={false}
      onCancel={[Function]}
      onConfirm={[Function]}
      onDateChange={[Function]}
      onHideAfterConfirm={[Function]}
      reactNativeModalPropsIOS={Object {}}
      titleIOS="Pick a date"
    />
    <CustomDatePickerIOS
      cancelTextIOS="Cancel"
      confirmTextIOS="Confirm"
      date={2019-08-13T14:43:38.802Z}
      dismissOnBackdropPressIOS={true}
      hideTitleContainerIOS={false}
      isVisible={false}
      mode="date"
      neverDisableConfirmIOS={false}
      onCancel={[Function]}
      onConfirm={[Function]}
      onDateChange={[Function]}
      onHideAfterConfirm={[Function]}
      reactNativeModalPropsIOS={Object {}}
      titleIOS="Pick a date"
    />
  </View>
  <TouchableHighlight
    activeOpacity={0.85}
    delayPressOut={100}
    disabled={true}
    onPress={[Function]}
    style={
      Object {
        "alignItems": "center",
        "alignSelf": "stretch",
        "backgroundColor": "#48BBEC",
        "borderColor": "#48BBEC",
        "borderRadius": 5,
        "height": 50,
        "justifyContent": "center",
        "margin": 10,
      }
    }
    underlayColor="black"
  >
    <Text
      style={
        Object {
          "color": "#fff",
          "fontSize": 18,
        }
      }
    >
      New Reservation
    </Text>
  </TouchableHighlight>
</View>
`;
